[WIP] BertAbs summarization

rlouf opened the pull request:
This PR builds on the encoder-decoder mechanism to do abstractive summarizaton. Contributions:

- A BeamSearch class that takes any `PreTranedEncoderDecoder` as an input;
- A script `run_summarization.py` that allows to pre-train the model and generate summaries.

Note that to save the checkpoints I had to add a parameter in the `save_pretrained_model` method, but I am not sure this is the best solution.

hlums commented:
@rlouf Thanks for the the PR!
I'm trying to run run_summarization.py with the following command, but got an error. Any idea why?
I have unzipped the cnn and dailymail folders in my data directory. 

```
python examples/run_summarization.py \
        --data_dir /data/home/hlu/notebooks/summarization/data \
        --output_dir ./summarization_output \
        --do_train True \
        --model_name_or_path bert-base-cased \
        --num_train_epochs 1 \
```

W1108 21:29:27.643532 139837444753152 run_summarization.py:488] Process rank: 0, device: cuda, n_gpu: 1, distributed training: False, 16-bits training: False
I1108 21:29:27.643716 139837444753152 run_summarization.py:491] Training/evaluation parameters Namespace(data_dir='/data/home/hlu/notebooks/summarization/data', device=device(type='cuda'), do_evaluate=False, do_overwrite_output_dir=False, do_train=True, gradient_accumulation_steps=1, max_grad_norm=1.0, max_steps=-1, model_name_or_path='bert-base-cased', model_type='bert', n_gpu=1, num_train_epochs=1, output_dir='./summarization_output', per_gpu_eval_batch_size=4, per_gpu_train_batch_size=4, seed=42, to_cpu=False)
I1108 21:30:40.805522 139837444753152 run_summarization.py:198] ***** Running training *****
I1108 21:30:40.805693 139837444753152 run_summarization.py:199]   Num examples = 312085
I1108 21:30:40.806245 139837444753152 run_summarization.py:200]   Num Epochs = 1
I1108 21:30:40.806311 139837444753152 run_summarization.py:201]   Instantaneous batch size per GPU = 4
I1108 21:30:40.806385 139837444753152 run_summarization.py:204]   Total train batch size (w. parallel, distributed & accumulation) = 4
I1108 21:30:40.806446 139837444753152 run_summarization.py:207]   Gradient Accumulation steps = 1
I1108 21:30:40.806501 139837444753152 run_summarization.py:208]   Total optimization steps = 78022
Traceback (most recent call last):
  File "examples/run_summarization.py", line 545, in <module>
    main()
  File "examples/run_summarization.py", line 497, in main
    global_step, tr_loss = train(args, model, tokenizer)
  File "examples/run_summarization.py", line 236, in train
    decoder_lm_labels=lm_labels,
  File "/data/anaconda/envs/nlp_gpu/lib/python3.6/site-packages/torch/nn/modules/module.py", line 547, in __call__
    result = self.forward(*input, **kwargs)
  File "/data/home/hlu/notebooks/huggingface/transformers/transformers/modeling_encoder_decoder.py", line 240, in forward
    decoder_outputs = self.decoder(decoder_input_ids, **kwargs_decoder)
  File "/data/anaconda/envs/nlp_gpu/lib/python3.6/site-packages/torch/nn/modules/module.py", line 547, in __call__
    result = self.forward(*input, **kwargs)
  File "/data/home/hlu/notebooks/huggingface/transformers/transformers/modeling_bert.py", line 865, in forward
    encoder_attention_mask=encoder_attention_mask)
  File "/data/anaconda/envs/nlp_gpu/lib/python3.6/site-packages/torch/nn/modules/module.py", line 547, in __call__
    result = self.forward(*input, **kwargs)
  File "/data/home/hlu/notebooks/huggingface/transformers/transformers/modeling_bert.py", line 677, in forward
    extended_attention_mask = causal_mask[:, None, :, :] * attention_mask[:, None, None, :]
RuntimeError: expected device cuda:0 and dtype Long but got device cuda:0 and dtype Bool

rlouf commented:
Hi! Thank you for raising the issue. I think it comes from the following lines in `modeling_bert.py`:

```python
causal_mask = seq_ids[None, None, :].repeat(batch_size, seq_length, 1) <= seq_ids[None, :, None]
extended_attention_mask = causal_mask[:, None, :, :] * attention_mask[:, None, None, :]
```

`causal_mask` as defined is a boolean tensor, while the `attention_mask` being passed is of type `long`. I am very confused because I never got the error when I tried locally. I will fix the issue asap. In the meantime, would you mind executing the following code in your environment and pasting the result here?

```python
import platform; print("Platform", platform.platform())
import sys; print("Python", sys.version)
import torch; print("PyTorch", torch.__version__)
```

hlums commented:
@rlouf Thanks! Please see the environment info below.
>>> import platform; print("Platform", platform.platform())
Platform Linux-4.15.0-1061-azure-x86_64-with-debian-stretch-sid
>>> import sys; print("Python", sys.version)
Python 3.6.8 |Anaconda, Inc.| (default, Dec 30 2018, 01:22:34) 
[GCC 7.3.0]
>>> import torch; print("Pytorch", torch.__version__)
Pytorch 1.2.0

rlouf commented:
Thank you! I’m working with PyTorch 1.3.0, and they introduced type promotion in this release (https://github.com/pytorch/pytorch/releases) which surely explains why it worked for me but not for you. I will work on a fix this week, in the meantime you can simply convert causal mask manually to long, or upgrade to pytorch 1.3 if that is possible.

julien-c commented on examples/model_bertabs.py:
```diff
@@ -0,0 +1,719 @@
+                                     attention_probs_dropout_prob=args.enc_dropout)
+            self.encoder.model = BertModel(bert_config)
+
+        if(args.max_pos > 512):
```
whitespace

rlouf replied:
Probably. I’m not changing the authors my code formatting for now, as it makes the comparison easier. Will run everything through `black` before merging though.

julien-c commented on examples/model_bertabs.py:
```diff
@@ -0,0 +1,719 @@
+
+    def forward(self, emb, step=None):
+        emb = emb * math.sqrt(self.dim)
+        if (step):
```
unneeded parens?

rlouf replied:
Indeed, thanks. There were a lot of those in the original code. Better safe than sorry?

rlouf commented:
Here is the latest on summarization:

- I added utilities to load the CNN/DailyMail dataset, to separate stories from summaries. The utilities are tested;
- I wrapped the BertAbs model to be able to use it with the library's usual API `BertAbs.from_pretrained`. I uploaded the finetuned weights and configuration on S3 and tested that model loads properly;
- I also wrapped the sequence generation so we can generate summaries from any text using Beam Search;
- I removed my previous attempt at Beam Search and saved it for later;
- I added a small fix in `modeling_bert.py` for users that use PyTorch v < 1.30. We convert the causal mask explicitly from `bool` to `long` since type promotion was only introduce in PyTorch 1.3

- [ ] Update the examples docs

I squashed my commits and rebased on the repo's master branch.

julien-c commented on examples/summarization/run_summarization.py:
```diff
@@ -0,0 +1,348 @@
+    return True
+
+
+def maybe_create_output_dir(path):
```
if the script is Python 3 only you can use `os.makedirs(exist_ok=True)`

rlouf replied:
[ah!](https://m.youtube.com/watch?v=XE6YaLtctcI)

julien-c commented on requirements.txt:
```diff
@@ -9,4 +9,7 @@ regex
+sacremoses
+# For ROUGE
+nltk
+py-rouge
```
I think we just add an other `requirements.txt` to the example subfolder

rlouf replied:
Ok, I agree it’s cleaner
